util/design/lib/OtpMemImg.py

```diff
@@ -413,6 +421,52 @@ def streamout_partition(self, part):
         # in the memory hex file.
         return data, annotation
 
+    def validate_data_perm(self, data_perm):
+        '''Validate data permutation option'''
+
+        # Byte aligned total width after adding ECC bits
+        total_bytelen = (
+            (self.lc_state.config['secded']['data_width'] +
+             self.lc_state.config['secded']['ecc_width'] + 7) // 8)
+        total_bitlen = total_bytelen * 8
+
+        # Default mapping
+        if not data_perm:
+            self.data_perm = list(range(total_bitlen))
+
+        # Parse custom mapping
+        else:
+            # Check the format first
+            pattern = r'^((?:\[[0-9]+:[0-9]+\])+(?:,\[[0-9]+:[0-9]+\])*)'
+            match = re.fullmatch(pattern, data_perm)
+            if match is None:
+                raise RuntimeError('Data permutation "{}" has invalid format'
+                                   .format(data_perm))
+            # Expand the ranges
+            self.data_perm = []
+            groups = match.groups()
+            for group in groups[0].split(','):
+                k1, k0 = [int(x) for x in group[1:-1].split(':')]
+                if k1 > k0:
+                    self.data_perm = list(range(k0, k1 + 1)) + self.data_perm
+                else:
+                    self.data_perm = list(range(k0, k1 - 1, -1)) + self.data_perm
+
+        # Check for bijectivity
+        if len(self.data_perm) != total_bitlen:
+            raise RuntimeError('Data permutation "{}" is not bijective'
+                               .format(data_perm))
+        accounting = dict()
+        for k in self.data_perm:
+            if k >= total_bitlen:
+                raise RuntimeError('Data permutation "{}" is not bijective'
+                                   .format(data_perm))
+            accounting.update({k: True})
+
+        if len(accounting) != total_bitlen:
+            raise RuntimeError('Data permutation "{}" is not bijective'
+                               .format(data_perm))
```

Comment:
It might be nice to provide slightly more informative messages for the different error cases here. I also think think the last check could be simpler: can't you just check that `len(set(self.data_perm)) == total_bitlen`? It will never be greater (because they are all non-negative numbers less than `total_bitlen`), but duplicates will mean the set is smaller.

Reply:
I've added some more description to differentiate between these error messages.
As for the last check: yes, a set conversion would be much more concise. I'll change that.